### PR TITLE
[examples] Add high_bias_absorb in CLE

### PR DESCRIPTION
**Pull request by zk1998**

### CODE REVIEW

File: examples/quantization/rep_quantization/README_zh-CN.md [hunk 1]
```diff
@@ -6,24 +6,41 @@
...
+```python
+# 典型的MobileOne进行CLE的示例
+# 进行BN_Restore之前需要指定需要进行bn恢复的卷积，否则默认会在所有卷积后都添加一个BN。
+layers_fused_bn = [name for name, mod in model.named_modules() if isinstance(mod, torch.nn.Conv2d) and 'reparam' in name]
+# MobileOne部署模型BN已经被融合到Conv中，进行BN恢复可以确保正常HBA。
+model = model_restore_bn(model, device, calibrate, context, layers_fused_bn)
+# 进行CLE，得到量化友好的MobileOne模型。
+model = cross_layer_equalize(model, dummy_input, device)
+```
+5. 在使用CLE的实际过程中，发现部分卷积的卷积核对应的特征输出会被放大10e8倍导致激活值量化失败，可以通过控制参数`threshold`来避免激活值量化失效。此外，单独进行一次CLE时，权重的均匀化程度有时还不能保证，可以尝试多次进行CLE进一步提升量化精度。
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/README_zh-CN.md:45]: 在使用的CLE的过程中，发现由于对weight的量化参数的调整，导致卷积部分通道的激活值被放大xxx，从而无法正常量化，在这种情况下，可以使用xxx

此外，若只进行一次CLE，权重的量化参数仍可能不均匀，导致较大的量化误差，可以尝试多次进行CLE进一步提升量化精度。

File: examples/quantization/rep_quantization/README_zh-CN.md [hunk 2]
```diff
@@ -6,24 +6,41 @@
...
+6. 可选操作：在量化时训练准备阶段，对rep类模型进行CLE变换和BN重建之后，可以对浮点模型进行finetune训练，使模型量化训练更加稳定。
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/README_zh-CN.md:46]: 在开始QAT之前，对RepVGG-like的模型。。。。

File: examples/quantization/rep_quantization/rep_ptq.py [hunk 3]
```diff
@@ -6,34 +6,39 @@
...
 
 
 def main_worker(args):
     # Provide a viable input for the model
     dummy_input = torch.rand((1, 3, 224, 224))
 
-    # We use BN_fused MobileNetV1 to simulate MobileOne_deploy which be re-parameterized.
-    # You can directly use the re-parameterized_to_deploy model.
-    model = mobilenet_fused_bn(dummy_input)
+    # Provide your pretrain model
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/rep_ptq.py:26]: pretrained

File: examples/quantization/rep_quantization/rep_ptq.py [hunk 4]
```diff
@@ -6,34 +6,39 @@
...
+    model = Mobilenet()
+    model.load_state_dict(torch.load(DEFAULT_STATE_DICT))
+
+    # Define the train_related context
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/rep_ptq.py:30]: training-related

File: examples/quantization/rep_quantization/rep_ptq.py [hunk 5]
```diff
@@ -6,34 +6,39 @@
...
+    device = get_device()
+    context = DLContext()
+    context.device = device
+    context.train_loader, context.val_loader = get_dataloader(args.data_path, 224, args.batch_size, args.workers)
 
-    # Do CLE(Optional), if weight of conv_bn_fused has some outliers which is hard to quantize, considering trying CLE.
-    cross_layer_equalize(model, dummy_input)
+    # if your pretrained model is Rep_style_deploy, use the following line to do bn restore to do High Bias Absorb.
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/rep_ptq.py:36]: 首字母大写
Rep_style_deploy是啥？RepVGG-like
to do xxx to do yyy -> to do xxx and yyy

File: examples/quantization/rep_quantization/rep_ptq.py [hunk 6]
```diff
@@ -6,34 +6,39 @@
...
+    # model = model_restore_bn(model, device, calibrate, context)
+    # Do CLE, if weight has some outliers which is hard to quantize, considering trying CLE.
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/rep_ptq.py:38]: 别再用DO了啊，Apply 或者 Perform也能用啊

File: examples/quantization/rep_quantization/rep_ptq.py [hunk 7]
```diff
@@ -6,34 +6,39 @@
...
+    model = cross_layer_equalize(model, dummy_input, device)
 
-    # Continue to do ptq.
+    # Now you get a model whose weights and activations is easy to quantize, continue to do ptq.
```
Comment (peterjc123) [inline examples/quantization/rep_quantization/rep_ptq.py:41]: is -> are
Now -> Now that
do的问题

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 8]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
+        processed_rules = load_processed_ptq_rules()
+        processed_rules = {nn.BatchNorm2d: processed_rules[nn.BatchNorm2d]}
+        is_fusable = functools.partial(quantizer.is_fusable, current_rules=processed_rules, graph=graph)
+        graph.filter_forward_nodes(is_fusable, custom_data, reverse=True)
+        quant_list = custom_data[0]
+        for quant_nodes in quant_list:
+            if isinstance(getattr(graph.module, quant_nodes[1]), nn.BatchNorm2d):
+                bn_cur = getattr(graph.module, quant_nodes[1])
+            torch_q.fuse_modules(graph.module, quant_nodes, inplace=True)
+            setattr(getattr(graph.module, quant_nodes[0]), 'fused_bn', bn_cur)
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:347]: 最好别直接打在module上面，就算一定要打也不要用这个名字，可以前面加个下划线之类的，避免冲突

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 9]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
+    c = pre_bn.bias - 3 * torch.abs(pre_bn.weight)
+    zero = torch.zeros_like(c)
+    c = torch.where(c < 0, zero, c).to(torch.float)
+    cur_weight = cur_conv.weight.data
+    # sum along 3rd and 4rd aixs
+    reduced_weight = cur_weight.sum(dim=[2, 3])
+    if reduced_weight.shape[1] == 1:
+        # for dw conv
+        reduced_weight = reduced_weight.reshape(-1)
+        bias_correct = reduced_weight * c
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:254]: 注意no_grad，或者用.detach()

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 10]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
+    """
+    cle_model.to(device)
+    origin_model = copy.deepcopy(cle_model)
+    bias_absorb_(cle_model, layer_groups, origin_model)
+    return origin_model
+
+
+def model_fuse_bn(model: nn.Module, dummy_input):
+    """fuse bn to conv inplace, and attach the origin bn to fused conv with attr:`fused_bn`"""
+    with model_tracer() and torch.no_grad():
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:325]: 虽然可以省掉一行，但是这取决于前面这个条件要返回true，考虑用with (A,B): 这样返回啥都行了

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 11]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
 
     Args:
         model: The bn of model should be fused into conv.
         dummy_input (torch.tensor): A viable input for the model.
-        threshold: (Optional) Default to be 1000, used to prevent unquantifiable anomalies in the output of inter conv.
+        threshold: Default to be 1000, used to prevent unquantifiable anomalies in the output of inter conv.
+    Returns:
+        typing.Tuple[List, nn.Module], layers groups and model after CLE.
     """
+    with torch.no_grad() and model_tracer():
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:169]: 一样

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 12]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
+
+        stat_we = model.state_dict()
+        for k, v in stat_we.items():
+            p, mod = cur_graph.get_submodule_with_parent_from_name(k)
+            if isinstance(mod, torch.nn.Conv2d):
+                if 'weight' in k:
+                    after_max = p.abs().max()
+                elif 'bias' in k:
+                    after_max = p.max()
+                log.info(f'{k}: {param[k].data.item():.5f} -> {after_max.data.item():.5f}')
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:191]: 只有after_max有效才打吧，否则打出很多无关的东西

Comment (peterjc123) [reply tinynn/graph/quantization/cross_layer_equalization.py:191]: 而且最好是endswith('.weight')这样，否则当中有个叫weight的什么东西直接不对了

File: tinynn/graph/quantization/cross_layer_equalization.py [hunk 13]
```diff
@@ -149,49 +144,243 @@ def _weight_equal_helper(cls, threshold):
...
+        threshold: Default to be 1000, used to prevent unquantifiable anomalies in the output of inter conv.
+    Returns:
+        typing.Tuple[List, nn.Module], layers groups and model after CLE.
     """
+    with torch.no_grad() and model_tracer():
+        cur_graph = trace(model, dummy_input)
+        param = {}
+        for k, v in model.state_dict().items():
+            p, _ = cur_graph.get_submodule_with_parent_from_name(k)
+            if 'weight' in k:
```
Comment (peterjc123) [inline tinynn/graph/quantization/cross_layer_equalization.py:174]: 一样

**Review by peterjc123 — CHANGES_REQUESTED** [review]
需要优化一下

**Comment by zk1998** [comment]
您好，我已经收到您的邮件，我会尽快查看并在第一时间回复您。

**Review by peterjc123 — CHANGES_REQUESTED** [review]

**Review by peterjc123 — APPROVED** [review]

**Comment by zk1998** [comment]
您好，我已经收到您的邮件，我会尽快查看并在第一时间回复您。
